File: Datasets/TrickyBugs/GenProgs/dpp_generated_progs_python/p03053/p03053_num7_parsed.py
import queue

# Function to check if the given cell (i, j) is a valid cell or not
def isValid(i, j, H, W):
    return i >= 0 and i < H and j >= 0 and j < W

# Function to perform the operation and count the number of operations
def countOperations(grid, H, W):
    # Initialize a visited array to keep track of visited cells
    visited = [[False for _ in range(W)] for _ in range(H)]
    
    # Initialize a queue for BFS traversal
    q = queue.Queue()
    
    # Initialize count to keep track of number of operations
    count = 0
    
    # Traverse the grid and find the black square(s) to start the operation
    for i in range(H):
        for j in range(W):
            if grid[i][j] == '#':
                q.put((i, j))
                visited[i][j] = True
    
    # BFS traversal to perform the operation and count the number of operations
    while not q.empty():
        size = q.qsize()
        
        while size > 0:
            size -= 1
            curr = q.get()
            row, col = curr[0], curr[1]
            
            # Check all four adjacent cells
            for k in range(4):
                x = row + dx[k]
                y = col + dy[k]
                
                if isValid(x, y, H, W) and not visited[x][y] and grid[x][y] == '.':
                    q.put((x, y))
                    visited[x][y] = True
        
        # Increment the count after each BFS traversal
        count += 1
    
    # Decrease the count by 1, as the last BFS traversal does not change anything
    count -= 1
    
    return count

# Read the input
H, W = map(int, input().split())

# Initialize the grid
grid = []
for _ in range(H):
    grid.append(list(input()))

# Define the four directions
dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

# Call the function to count the number of operations
num_operations = countOperations(grid, H, W)

# Print the result
print(num_operations)

File: Datasets/TrickyBugs/GenProgs/dpp_generated_progs_cpp/p03053/p03053_num7_parsed.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

// Function to check if a given position is valid or not
bool isValid(int x, int y, int row, int col) {
    return (x >= 0 && x < row && y >= 0 && y < col);
}

// Function to perform BFS traversal
int performOperations(vector<vector<char>>& grid) {
    int row = grid.size();
    int col = grid[0].size();
    int operations = 0;
    
    // Define directions for movement in the grid
    int dx[] = {-1, 0, 1, 0};
    int dy[] = {0, 1, 0, -1};
    
    // Create a visited array to keep track of visited squares
    vector<vector<bool>> visited(row, vector<bool>(col, false));
    
    // Create a queue to perform BFS
    queue<pair<int, int>> q;
    
    // Find the initial black squares and add them to the queue
    for (int i = 0; i < row; i++) {
        for (int j = 0; j < col; j++) {
            if (grid[i][j] == '#') {
                q.push({i, j});
                visited[i][j] = true;
            }
        }
    }
    
    // Perform BFS traversal
    while (!q.empty()) {
        int size = q.size();
        
        for (int i = 0; i < size; i++) {
            int x = q.front().first;
            int y = q.front().second;
            q.pop();
            
            // Check all adjacent squares
            for (int k = 0; k < 4; k++) {
                int nx = x + dx[k];
                int ny = y + dy[k];
                
                // If adjacent square is valid and white, change it to black and add to queue
                if (isValid(nx, ny, row, col) && !visited[nx][ny] && grid[nx][ny] == '.') {
                    q.push({nx, ny});
                    visited[nx][ny] = true;
                    grid[nx][ny] = '#';
                }
            }
        }
        
        // Increment the number of operations performed
        operations++;
    }
    
    return operations - 1; // Subtract 1 as the initial black squares are already counted
}

int main() {
    int H, W;
    cin >> H >> W;
    
    vector<vector<char>> grid(H, vector<char>(W));
    
    for (int i = 0; i < H; i++) {
        for (int j = 0; j < W; j++) {
            cin >> grid[i][j];
        }
    }
    
    // Call the function to perform operations and print the result
    cout << performOperations(grid) << endl;
    
    return 0;
}